### ISSUE TITLE
wxCrafter issue with PNG wxBitmap button

### ISSUE THREAD

**Issue by tester0077** [eranif/codelite#2830]
Running Win 10, 64-bit. - all updates  MSVC 2019, wxWidgets 3.1.4
Created a dialog with wxCrafter 2.9
Have some bitmap buttons, all display ok with the expected PNG file.
Added one more wxBitmap button, picked a new PNG 16x16 and while it shows as expected in the wxCrafter, when the files from wxCrafter are compiled into the project, the button does not display at all and my log tells me:
11:55:03: Error: XRC error: XRC resource "edit-paste" (class "wxBitmap") not found

The icon came from the tango theme 0.8.90 16x16 actions set
I can select other icons in its place and some will display, other will not.
I have checked the failing icon with pngcheck -vv 
```
 D:\pkg\wx\MSVC2019\_3.1.4_2019\wxET\rc> D:\Downloads\Graphics\PNG\pngcheck-3.0.3-win32\pngcheck.win64.exe -vv .\edit-paste.png
File: .\edit-paste.png (546 bytes)
  chunk IHDR at offset 0x0000c, length 13
    16 x 16 image, 8-bit palette, non-interlaced
  chunk PLTE at offset 0x00025, length 273: 91 palette entries
  chunk tRNS at offset 0x00142, length 23: 23 transparency entries
  chunk IDAT at offset 0x00165, length 169
    zlib: deflated, 32K window, fast compression
    row filters (0 none, 1 sub, 2 up, 3 avg, 4 paeth):
      0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 (16 out of 16)
  chunk IEND at offset 0x0021a, length 0
No errors detected in .\edit-paste.png (5 chunks, -113.3% compression).
```

![edit-paste](https://user-images.githubusercontent.com/9053628/133939643-b2a83372-4d8d-40a2-ac92-6c60fda1ee50.png)

Not a show stopper, but still curious how to fix it or work around/ possible causes

**Comment by tester0077** [comment]
It definitely is a problem with specific PNGs, as the same issue arises when I place the same PNG into a AUI tool bar.

Still unsure whether it is an issue with wxCrafter or wxWIdgets

**Comment by tester0077** [comment]
After more chasing down possible clues, I am convinced the problem is somewhere within
the line
m_bmpButtonEdit = new wxBitmapButton(m_panel441, wxID_WS_EDIT_ENTRY, wxXmlResource::Get()->LoadBitmap(wxT("Modify_16x16")), wxDefaultPosition, wxDLG_UNIT(m_panel441, wxSize(-1,-1)), wxBU_AUTODRAW);

or with the way wxCrafter creates the resource it tries to load.
I can load the PNG the resource is derived from by wxCrafter directly as a bitmap
SO I don't see any issue with the specific PNG from the wxWindows side

**Comment by tester0077** [comment]
Have build a minimal test example and posted it on the wxWidgets forum for help.

**Comment by tester0077** [comment]
There may well be issues with my test code, because I am not familiar enough with the wxCrafter way of loading images.
Meanwhile, to be able to carry on development of my app, I have resorted to a work-around, asking wxCrafter to use images from wxArtProvider

**Comment by termi3** [comment]
Probably the same problem for me - I succeed to solve it, after I have included  wxInitAllImageHandlers(); just after the block 

if ( !bBitmapLoaded ) {
        // We need to initialise the default bitmap handler
        wxXmlResource::Get()->AddHandler(new wxBitmapXmlHandler);
        wxC1F0BInitBitmapResources();
        bBitmapLoaded = true;
    }  
> here   wxInitAllImageHandlers();
in the generated file. Then everything  works fine, but this file is generated by wxCrafter.  So anytime I change something in the GUI have to add it.

**Comment by dghart** [comment]
> but this file is generated by wxCrafter. So anytime I change something in the GUI have to add it.

Why not instead put wxInitAllImageHandlers() in your program's MyApp::OnInit, before the wxC code is loaded?

Regards,

David

**Comment by tester0077** [comment]
In the app which shows this issue, ::wxInitAllImageHandlers(); is called in OnInit(), but the problem shows in a dialog, with its GUI also constructed using wxCrafter, but in a separate sub-project, built into to a separate static library, linked to the main project.
For this case, adding ::wxInitAllImageHandlers(); in the dialog base class ctor does not fix the issue. I have tried to added the call to that function before, after and as part of the `if ( !bBitmapLoaded ) {` block, none of these options work, nor does including it in the derived dialog class ctor.
There is something unexpected in some PNG and wxWindows chokes on it.
For my work, I have retreated to some wxArtProvider equivalents, at least for the time being
The attached PNG is one which shows the issue in my case.
![Modify_16x16](https://user-images.githubusercontent.com/9053628/134946090-5e4c8d3f-99c5-4bed-906d-59ce4152847f.png)

**Comment by dghart** [comment]
Your png works for me, after I hacked it into one of CodeLite's template projects: 'XRC frame'.
![wxMiniApp](https://user-images.githubusercontent.com/6452587/134958019-12020415-554b-47e5-b5ba-b97245c4a174.png)

**Comment by tester0077** [comment]
Interesting. Either the XRC data it generates and tries to load in my case differs form the plain XRC case that works for you, or the whole process differs.
Any thoughts on how to move forward?
As I said, I am using wxCrafter standalone and have used its wxCrafter plugin to confirm that it makes no difference to this issue, in my case, but other than that, being used to the Windows environment, IDE and debugger, I have used Codelite only once several years ago in an effort to port one of my  wxWidgets app to the Linux world. The  port worked well enough, but never went beyond the proof of concept stage.
In the current app, it "would be nice if" (WBNI) it worked for the PNGs which cause me trouble, but there are always other ways to work around it. :-)

**Comment by dghart** [comment]
Out of curiosity if nothing else, what happens if you use e.g. ImageMagick to convert the failing png to e.g. a jpg? Does that then load? If so, try converting it back to a png and retry. If that also works, you could also try converting the original png into another png e.g. shrink it by 1%.

**Comment by tester0077** [comment]
I had originally tried some other formats without success,
I have now repeated some of these tests with an XMP derived from a failing PNG - converted, I believe by ImageBlocks.
Similarly, I have used an ico file of another failing PNG - also via ImageBl;ocks
Today I have used the original 24x24 version of the PNG obtained from the collection
D:\Downloads\Icons\Aha-Soft-24x24-free-application-icons\png\24x24
I have used the latest Gimp to scale that to 20x20, then to 16x16 and created a 16x16 JPG from that
All fail with the respective error:
14:06:00: Error: XRC error: XRC resource "clipboardArrow" (class "wxBitmap") not found  <<XPM
14:06:00: Error: XRC error: XRC resource "edit_paste" (class "wxBitmap") not found           << ICO
14:06:00: Error: XRC error: XRC resource "Modify" (class "wxBitmap") not found    << original PNG from Ahasoft 24x24
14:06:00: Error: XRC error: XRC resource "Modify20x20" (class "wxBitmap") not found  << scaled by GIMP  to 20x20 -> PNG
14:06:00: Error: XRC error: XRC resource "Modify16x16-2" (class "wxBitmap") not found << scaled by GIMP  to 16x16 -> JPG

Not at all sure what that might tell you :-)
As GIMP has so many options for export, I was unsure which to use, so I simply stuck with the defaults.

**Comment by dghart** [comment]
> Not at all sure what that might tell you :-)

More 'confuse' than 'tell' ;)

Let me just check I understand the situation correctly. You have a dialog that contains 'n' bitmaps. Most display correctly, but 1 or more fail to load. That dialog is a loaded from a statically-linked library.

I can't think why wx would fail to load one png from a static-linked lib, but successfully load sibling pngs, if wxInitAllImageHandlers() were somehow invisible inside the lib. And the minimal sample that you posted on wxForum (presumably) produced the error for you, though it displays fine for me (debian 10), though it understandably chokes on modifywxCrafter.xml as it isn't an image file. Do you have other platforms available on which you can test it?

Finally, scraping the bottom of the barrel: in your first post you wrote
>Added one more wxBitmap button...

What if you change the order by swapping an earlier 'good' png for the failing one; does the same one still fail, although it's loaded sooner.

**Comment by tester0077** [comment]
After getting away from the problem for a while, I have revisited the test project and my last tests with my own project using wxCrafter and its output. 
The test project does not rely on wxCrafter or any GUI base code created by it. In that example I can display the 'failing' PNGs without issue. The problem with the test project all stemmed from my trying to use the 'bad' data produced by wxCrafter for my 'real project and it is only this part for which it produces any errors.
IOW, currently, the project, built without the GUI builder, does display the images correctly as long as I do not use wxCrafter generated data.

Testing on a different platform would require that I resurrect the Linux Mint environment I had last used several years ago. That does not seem feasible right now.

My own project, with its GUI built by wxCrafter, does not display the same and several other PNGs and instead displays an error message in my log.  But this is only the case for some specific PNGs and it seems to not matter whether I convert those to ICO, XPM, or JPG. Nor does the sequence in loading these or other 'good' images matter. 
My conclusion is still that this somehow is a problem within the wxCrafter code & workflow.

This may well be specific to Windows, but at present I do not have the means to follow this issue down that path.

**Comment by dghart** [comment]
I see. However I don't have Windows here and, anyway, I can't reproduce/debug the problem without a failing workspace/project to test.

If you have the time and inclination to produce and post such a project, I'm happy to try to debug it.

**Comment by tester0077** [comment]
Not sure I understand. I am assuming you used the plain wxWidgets based sample I posted.
Or did you actually create/reproduce the GUI using Codelite/wxCrafter?

**Comment by dghart** [comment]
>I am assuming you used the plain wxWidgets based sample I posted.

That's right. It works for me: the first 2 buttons, Modify_16x16.png and Modify_16x16.xpm, display correctly. The third, "D:/MSVC2019/wxWidgets-3.1.4/samples/minimalPng/modifywxCrafter.xml and modifywxCrafter.xml, fail as expected since they are trying to load something that isn't a bitmap (see doublemax's wxforum explanation).

I presumed that, for you, one or both of the first 2 bitmaps were failing. Was I wrong? But either way, it's behaving as expected here.

**Comment by tester0077** [comment]
At the time I posted the example it was the third button which gave me the problem as it does for you. As well, at that time I did not understand the way that wxCrafter implemented its button graphics - not sure I really understand even now - and DoubleMax's explanation certainly helped.
In any case, my conclusion is that the problem is within the wxCrafter portion of the code.
And since I have found a work-around using wxArtProvider, I won't pursue this myself but will leave the issue open, hoping that it will catch the eye of someone able to resolve it.
Thank you for your help with this; it has helped clear up some 'fog' in my thinking along the way

**Comment by dghart** [comment]
>Thank you for your help with this; it has helped clear up some 'fog' in my thinking along the way

I'm glad it has, but my understanding of the situation is still hazy. Is it only bitmaps loaded from those xml files that fail? If so, how do you get wxCrafter to generate those xml files? Have you been using 'Generate C++ code', 'Generate XRC', or something else?

**Comment by tester0077** [comment]
I have been using wxCrafter for so long now that I don't even think about what it really does :-)
Basically, I'll build the GUI elements into a frame or dialog, tell it what base name ( say xxxBase) I want it to use and then ask it to create the source files by clicking on the 'Generate Code" tool bar button
xxxBase.cpp, xxxBase.h, xxxBase_bitmaps.cpp & xxxBase_bitmaps.xrc 
Of these 4, only 3 are included as part of my project, the 4th, the XRC file, I simply ignore.
In the current case, as part of the layout, I have embedded some wxBitmap buttons and then assign the image file I want to be displayed. In most cases, I select a PNG and wxCrafter does all the rest. The image data is converted to an array which ends up in the file xxxBase_bitmaps.cpp as an array - which is what I had tried to load in my sample, using code I had copied/extracted from the wxCrafter generated files
So, yes, I have been generating C++ code, which forms the base classes for my frames, dialog etc.
The main issue here is that some PNGs embedded in the GUI using the usual work flow with wxCrafter work and display without any issue, the ones I have used in the example do not display at all.

**Comment by dghart** [comment]
>the XRC file, I simply ignore

Perhaps you have both of the Generate Code options ticked.
![GenerateCodeMenu](https://user-images.githubusercontent.com/6452587/135430665-2d966472-5a64-48ff-9b07-cbfb2e6b38da.png)

>The main issue here is that some PNGs embedded in the GUI using the usual work flow with wxCrafter work and display without any issue, the ones I have used in the example do not display at all.

I see. Well, if the first 2 bitmapsbuttons in the sample work for both of us, and the ones that are supposed to load from modifywxCrafter.xml fail for both of us, I can't see a way forward: I've no idea how the xml file is supposed to load, so I can't guess what might be going wrong.

At least you've found a work-around.

**Comment by tester0077** [comment]
![wxCrafterGenerateCode](https://user-images.githubusercontent.com/9053628/135480609-3879942f-7f93-40ce-901d-ce159bd451f5.jpg)
My version does not have 'Generate XRC ' checked, but does it anyhow. ;-)
The work-around is only  useful, because I have found an image that'll do for my purposes, at least for now, so I can carry on until it becomes a 'real' issue for this button, or for another where I can't find anything suitable.
FWIW, in testing this, I have found several other images from public domain icon sources, which show the same issue and what is worse, in my estimation, that even converting these to other formats does not seem to resolve the issue.
For the longest time I have had to 'work around' the warning ""iCCP: known incorrect sRGB profile"  and I still can't help but wonder where the current problem really lies. wxWidgets or wxCrafter?

The only way to resolve the issue myself, would be to download the full open-source project. build it and track down the cause.
That is far too steep a mountain, I am afraid, especially, since I barely know how to spell 'Linux', let alone feel comfortable debugging code built under Codelite, using gcc, along with its debugger, possibly using CMake to create an MSVC compatible solution file. 
It is far too dangerous for me to assume that I can do this work in the familiar Windows environment.

**Comment by tester0077** [comment]
Another wrinkle in this saga stopped me cold in my tracks for a bit, though I still have only come up with a work-around - again.
In this case, I was using a group of small bitmap buttons with PNG images in two wxCrafter dialogs for very similar functionality.
Each of the two dialogs is created from a separate wxCrafter 'project' and ends up as a separate statically linked library for the main project. One, the 'Options' dialog was created and incorporated first into the project.
Later on, the second dialog  'borrowed' the button group for a similar need - allowing the user to move lines up & down in a wxListCtrl, as well as to edit/add/delete lines in the control. 
In both cases I used the same resource PNGs and all worked well.
As the project progressed, I realized/decided that I no longer wanted/needed the group of buttons in the options dialog and dropped the GUI elements from the dialog and modified its source files accordingly.
To my big surprise, the buttons in the second dialog, whose code and wxCrafter project had not been touched - though it had been recompiled as part of a general 'clean' operation - disappeared as well.
The space these buttons occupied in the second dialog remained, but no bitmap, nor even an outline square, was visible. And, of course, nothing happened when one clicked where those buttons were supposed to be.
Backtracking and experimenting over the course of a day, showed that as long as I kept the button group somewhere in the fist - options - dialog, the buttons would also show and be usable in the second dialog. I could even hide and/or disable the buttons and their container in the fist dialog. 
To make this already long story somewhat shorter, I ended up removing the buttons from the first dialog and moving the library for the second dialog ahead of the one for the options dialog in the link step and my buttons appeared where they ought to be in the second dialog.
This will be my work-around for now, but I did want to record this for any one else who might run into a similar issue
